src/Orchard.Environment.Shell/Builders/ShellContainerFactory.cs

```diff
@@ -87,20 +86,24 @@ public IServiceProvider CreateContainer(ShellSettings settings, ShellBlueprint b
             var featureServiceCollections = new Dictionary<IFeatureInfo, ServiceCollection>();
 
             // Let any module add custom service descriptors to the tenant
-            foreach (var startup in moduleServiceProvider.GetServices<Microsoft.AspNetCore.Mvc.Modules.IStartup>())
+            foreach (var startup in moduleServiceProvider.GetServices<Microsoft.AspNetCore.Modules.IStartup>())
             {
                 var feature = blueprint.Dependencies.FirstOrDefault(x => x.Type == startup.GetType())?.Feature;
 
                 ServiceCollection featureServiceCollection;
                 ServiceCollection startupServices = new ServiceCollection();
+
                 if (!featureServiceCollections.TryGetValue(feature, out featureServiceCollection))
                 {
                     featureServiceCollections.Add(feature, featureServiceCollection = new ServiceCollection());
                 }
 
+                startupServices.Add(featureServiceCollection);
+                startupServices.Add(tenantServiceCollection);
                 startup.ConfigureServices(startupServices);
-                featureServiceCollection.Add(startupServices);
+
                 tenantServiceCollection.Add(startupServices);
+                tenantServiceCollection = new ServiceCollection().Add(tenantServiceCollection.Distinct());
```

Comment:
Are you sure? E.g about featureServiceCollection which has been added to featureServiceCollections it seems that here it stays empty. And featureServiceCollections is used later to register feature services in ITypeFeatureProvider.

Then, startupServices.Add(featureServiceCollection) seems to add an empty collection.
Then, maybe you need startupServices.Add(tenantServiceCollection) for other reasons.